### ISSUE TITLE
4.4.2 Repeated,unavoidable CTD during AI turn (modern/atomic era, Non-EUI, no other mods)

### ISSUE THREAD

**Issue by sofrondr** [LoneGazebo/Community-Patch-DLL#10644]
_1. Mod Version (X.Y.Z). Current Version: 4.4.2_
4.4.2 no-EUI, modpack for multiplayer made without any issues by me

_2. DirectX Version_

DX 10 & 11

_3. List of Other Mods_

No other mods

_4. Describe the Issue_

In this multiplayer game (I am host), clicking next turn results in a CTD for both me and the other one non-host. It happens during AI turns.

______________________________________________________________________________________
_5. Save Game From 1 Turn Before (ALWAYS ATTACH THIS IF POSSIBLE)_

[AutoSave_0310 AD-1880.zip](https://github.com/LoneGazebo/Community-Patch-DLL/files/14229245/AutoSave_0310.AD-1880.zip)

[Last 11 autosaves.zip](https://github.com/LoneGazebo/Community-Patch-DLL/files/14229246/Last.11.autosaves.zip), I can provide all 310 if needed.

_6. Logs (ALWAYS ATTACH THESE IF POSSIBLE)_

[LogsCTD20240209.zip](https://github.com/LoneGazebo/Community-Patch-DLL/files/14229247/LogsCTD20240209.zip)

_7. CvMiniDump.dmp File (ATTACH IF REPORTING A GAME CRASH)_

[CvMiniDump.zip](https://github.com/LoneGazebo/Community-Patch-DLL/files/14229249/CvMiniDump.zip)

_8. Steps to reproduce the Issue (Optional)_

Do whatever in the last turn, click Next Turn and the crash happens.

_9. Screenshots of the Issue (Optional)_

**Comment by sofrondr** [comment]
We have played two games this patch, and we were stopped around the same time by probably the same issue (my second report is #10626 , no replies/solution unfortunately).

**Comment by EliskaSvobodova** [comment]
I am the non-host, here are my files:
 
[CvMiniDump.zip](https://github.com/LoneGazebo/Community-Patch-DLL/files/14229287/CvMiniDump.zip)

[Logs20240209.zip](https://github.com/LoneGazebo/Community-Patch-DLL/files/14229285/Logs20240209.zip)

**Comment by ilteroi** [comment]
callstack

```
00 71d1f1dc 619a7ada     CvGameCore_Expansion2!CvPlayerEspionage::ProcessSpy+0x2ef [c:\repos\civ-vp\cvgamecoredll_expansion2\cvespionageclasses.cpp @ 589] 
01 71d1f224 6a333c3a     CvGameCore_Expansion2!CvPlayerEspionage::DoTurn+0x2a [c:\repos\civ-vp\cvgamecoredll_expansion2\cvespionageclasses.cpp @ 454] 
```

source:

```
							CvModEventCityChoiceInfo* pkEventChoiceInfo = GC.getCityEventChoiceInfo(pCityEspionage->GetCounterSpyFocus());
							if (!pkEventChoiceInfo->isSecretMission())
```

so, null pointer access. workaround is easy but it seems something is wrong in the database ...

**Comment by ilteroi** [comment]
try this build to continue your game

[CvGameCore_Expansion2-crash-fix.zip](https://github.com/LoneGazebo/Community-Patch-DLL/files/14232055/CvGameCore_Expansion2-crash-fix.zip)

**Comment by axatin** [comment]
Seems that a counterspy is in a city and has no focus, which shouldn't happen.

@sofrondr Please use this file instead, it also avoids the crash and prints some additional log messages. After having played the turn in which the game crashed before, please upload the log file "EspionageLog.csv".

[CvGameCore_Expansion2.zip](https://github.com/LoneGazebo/Community-Patch-DLL/files/14232814/CvGameCore_Expansion2.zip)

**Comment by Cyclopia31** [comment]
If I remember some of Seroperson's patches correctly, most of the spy system-related multiplayer CTDs he fixed were due to unnecessary network events being sent. (Example: https://github.com/LoneGazebo/Community-Patch-DLL/pull/9958). Maybe this is also the case here ?

**Comment by sofrondr** [comment]
@axatin Thank you for the patch, we added the file into the modpack (not making it again, that should not matter, right?) and we were able to progress 3 turns without CTD. We havent tested beyond that, but it looks ok.

> Seems that a counterspy is in a city and has no focus, which shouldn't happen.
> 
> @sofrondr Please use this file instead, it also avoids the crash and prints some additional log messages. After having played the turn in which the game crashed before, please upload the log file "EspionageLog.csv".
> 

Here is the log from host, those three turns (311-313, 311 is the one that was previously failing). Something about Chinese (the non-host player) spy not having focus. In Espion. UI in game, this spy has the first focus set (conducting counter-intelligence).
[EspionageLogHost.zip](https://github.com/LoneGazebo/Community-Patch-DLL/files/14236385/EspionageLogHost.zip)

We also get two warnings on turns 312,313 that there is sync issue on non-host PC.

**Comment by EliskaSvobodova** [comment]
Same log from these three turns from non-host:
[EspionageLogNonHost.zip](https://github.com/LoneGazebo/Community-Patch-DLL/files/14236439/EspionageLogNonHost.zip)

Thank you

**Comment by Cyclopia31** [comment]
> Seems that a counterspy is in a city and has no focus, which shouldn't happen.
> 
> @sofrondr Please use this file instead, it also avoids the crash and prints some additional log messages. After having played the turn in which the game crashed before, please upload the log file "EspionageLog.csv".
> 
> [CvGameCore_Expansion2.zip](https://github.com/LoneGazebo/Community-Patch-DLL/files/14232814/CvGameCore_Expansion2.zip)

Hi @axatin, is the fix integrated on the new patch? Thank you !

**Comment by axatin** [comment]
No, not yet. The file I gave you just avoids the crash, but it doesn‘t fix the underlying issue. I don‘t have access to my computer right now and can‘t fix it properly until beginning of March.

The problem is that when a human player selects a mission for their counterspy, that information is not transferred to the host. So, counterspies by humans never have an active mission in MP and that can cause crashes (and makes them useless even if the game doesn‘t crash). You can play with 4.5 as long as you don‘t use counterspies, or you can just turn espionage off for now. 

The fix for this will be savegame compatible.

**Comment by Cyclopia31** [comment]
Thank you for the detailed and quick answer !

**Comment by axatin** [comment]
Okay, I've made some changes and it would be great if some of you could test if it works and if counterspies are now multiplayer-compatible. I've attached some files, please use them in a 4.5-based multiplayer game and test: One player moves a counterspy to a city and selects the "counterspy network" mission, the other player sends a spy to the city. When the spy arrives, a notification should be given that a counterspy is in the city and network point generation should be reduced. Please test also if moving the counterspy out of the city or changing its focus works correctly.

The attached files:
- the dll, put it into (1) Community Patch
- EspionageChoicePopup.lua, put it into (1) Community Patch\Core Files\CoreLua
- EspionageOverview.lua and .xml, put them into (2) Vox Populi\Core Files\Overrides

[files.zip](https://github.com/LoneGazebo/Community-Patch-DLL/files/14487347/files.zip)

**Comment by Cyclopia31** [comment]
> Okay, I've made some changes and it would be great if some of you could test if it works and if counterspies are now multiplayer-compatible. I've attached some files, please use them in a 4.5-based multiplayer game and test: One player moves a counterspy to a city and selects the "counterspy network" mission, the other player sends a spy to the city. When the spy arrives, a notification should be given that a counterspy is in the city and network point generation should be reduced. Please test also if moving the counterspy out of the city or changing its focus works correctly.
> 
> The attached files:
> 
>     * the dll, put it into (1) Community Patch
> 
>     * EspionageChoicePopup.lua, put it into (1) Community Patch\Core Files\CoreLua
> 
>     * EspionageOverview.lua and .xml, put them into (2) Vox Populi\Core Files\Overrides
> 
> 
> [files.zip](https://github.com/LoneGazebo/Community-Patch-DLL/files/14487347/files.zip)

Hi Axatin, do we have to put these files before creating the modpack (so in the MOD folder) or after (so in the modpack) ? I guess it is before as there is an .xml file but I prefer to ask !

**Comment by Cyclopia31** [comment]
I tested a little bit and everything seem OK !

Just the message is incomplete: 
<img width="230" alt="Message" src="https://github.com/LoneGazebo/Community-Patch-DLL/assets/80354202/9538a6f2-21a0-40ed-a46c-602c6aa1d2f8">

But the counterspy works: 
<img width="399" alt="Spy Screen" src="https://github.com/LoneGazebo/Community-Patch-DLL/assets/80354202/0c181a5a-892e-40ba-861f-6f9d2d078cae">

**Comment by axatin** [comment]
Good! Then the fix will be added in the next version.

The incomplete message is an EUI bug unrelated to multiplayer, will be fixed as well.

**Comment by Cyclopia31** [comment]
I just tested the fix for about 20 turns but I'm pretty sure it works fine. It will be tested more when the next patch comes out officially with the fix :)
